File: Trees/del_node_bt.cpp
#include <iostream>
using namespace std;
#include <queue>

struct Node{
    int key;
    Node * left;
    Node * right;
};

Node* newNode(int key){
    struct Node* temp = new Node;
    temp->key = key;
    temp->left = NULL;
    temp->right = NULL;
    return temp;

};

/*Function to delete the deepest node 
in a given tree */
void inorder(Node* temp){
    if (!temp)
    return;
    inorder(temp->left);
    cout <<temp->key<<" ";
    inorder(temp->right);

}

void deldeepest(Node* root,Node* d_node){

    queue<struct Node*> q;
    q.push(root);
    //we perform level order traversal till last node
    struct Node* temp;
    while(!q.empty()){
        temp = q.front();
        q.pop();
        if (temp==d_node){
            temp = NULL;
            delete (d_node);
            return;
        }
        if (temp->right == d_node){
            temp = NULL;
            delete (d_node);
            return;
        }
        else(q.push(temp->right));
        if (temp->left == d_node){
            temp = NULL;
            delete(d_node);
            return;
        }else(q.push(temp->left));
        
    }

} 

Node* deletion(Node* root,int key){
    if (root ==NULL)
       return NULL;
    if (root->left ==NULL && root->right == NULL){
        if (root->key == key)
            return NULL;
        else
             return root;
    }
    queue<Node*> q;
    q.push(root);
    struct Node* temp;
    struct Node* key_node = NULL;
    while(!q.empty()){
        temp = q.front();
        q.pop();
        if (temp->key == key){
            key_node = temp;
        }
        if (temp->left)
            q.push(temp->left);
        if (temp->right)
            q.push(temp->right);
        
    }
    if (key_node !=NULL){
        int x =  temp->key;
        deldeepest(root,temp);
        key_node->key = x; 

    } 
    return root;
}

int main(){
    struct Node* root = newNode(10);
    root->left = newNode(11); 
    root->left->left = newNode(7); 
    root->left->right = newNode(12); 
    root->right = newNode(9); 
    root->right->left = newNode(15); 
    root->right->right = newNode(8);
    cout << "Inorder traversal before deletion : "; 
    inorder(root);
    cout << endl;
    root = deletion(root,12);
    cout << "Inorder traversal after deletion : "; 
    inorder(root);
    cout << endl;
    
    

}

File: Trees/del_node_bt.py
from collections import deque
class Node:
    def __init__(self,key):
        self.key = key
        self.left = None
        self.right = None


def inorder(root):
    curr = root
    s = []

    while True:
        while curr:
            s.append(curr)
            curr = curr.left
        if s:
            node = s.pop()
            print(node.key,end = ' ')
            curr = node.right
        else:
            return


def del_last_node(root,d_node):
    q = deque()
    q.append(root)
    while q:
        curr = q.popleft()
        if curr.left:
            if curr.left is d_node:
                curr.left = None
                return
            else:
                q.append(curr.left)
        if curr.right:
            if curr.right is d_node:
                curr.right = None
                return
            else:
                q.append(curr.right)

def del_node(root,key):
    q  = deque()
    key_node = None
    q.append(root)

    while q:
        curr  = q.popleft()
        if curr.key == key:
            key_node = curr

        if curr.left:
            q.append(curr.left)
        if curr.right:
            q.append(curr.right)
    if key_node:
        x = curr.key
        del_last_node(root,curr)
        key_node.key = x





            




if __name__ == '__main__':
    '''
         4
        / \
       3   5
      / \
     2   1
          \
           7
     '''
    root = Node(4)
    root.left = Node(3)
    root.left.left = Node(2)
    root.left.right = Node(1)
    root.left.right.right = Node(7)
    root.right = Node(5)
    print("The tree before the deletion:") 
    inorder(root) 
    key = 7
    del_node(root, key) 
    print() 
    print("The tree after the deletion;") 
    inorder(root) 